File: Data Structures & Algorithms/search-2d-matrix/submission-16.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int ROWS = matrix.length, COLS = matrix[0].length;
        int r = ROWS - 1, c = 0;

        while (c < COLS && r >= 0) {
            // staircase search, if too small move to the right
            if (matrix[r][c] < target) {
                c++;
            } else if (matrix[r][c] > target) { // if too large move up
                r--;
            } else {
                return true;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/binary-search/submission-16.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums)

        while l < r:
            m = l + (r - l) // 2

            if nums[m] > target:
                r = m
            else:
                l = m + 1
        
        return l - 1 if (l > 0 and nums[l - 1] == target) else -1
        